data-pipeline/src/telemetry/mod.rs

```diff
@@ -0,0 +1,568 @@
+    pub fn set_language(mut self, lang: &str) -> Self {
+        self.language = Some(lang.to_string());
+        self
+    }
+
+    /// Sets the language version for the telemetry client
+    pub fn set_language_version(mut self, version: &str) -> Self {
+        self.language_version = Some(version.to_string());
+        self
+    }
+
+    /// Sets the tracer version for the telemetry client
+    pub fn set_tracer_version(mut self, version: &str) -> Self {
+        self.tracer_version = Some(version.to_string());
+        self
+    }
+
+    /// Sets the url where the metrics will be sent.
+    pub fn set_url(mut self, url: &str) -> Self {
+        let _ = self
+            .config
+            .set_endpoint(ddcommon::Endpoint::from_slice(url));
+        self
+    }
+
+    /// Sets the heartbeat notification interval in millis.
+    pub fn set_hearbeat(mut self, interval: u64) -> Self {
+        if interval > 0 {
+            self.config.telemetry_hearbeat_interval = Duration::from_millis(interval);
+        }
+        self
+    }
+
+    /// Builds the telemetry client.
+    pub async fn build(self) -> Result<TelemetryClient, TraceExporterError> {
+        let (worker, handle) = TelemetryWorkerBuilder::new_fetch_host(
+            self.service_name.unwrap(),
+            self.language.unwrap(),
+            self.language_version.unwrap(),
+            self.tracer_version.unwrap(),
+        )
+        .spawn_with_config(self.config)
+        .await
+        .map_err(|_| TraceExporterError::Builder(BuilderErrorKind::InvalidTelemetryConfig))?;
+
+        Ok(TelemetryClient {
+            handle,
+            metrics: Metrics::new(&worker),
+            worker,
+        })
+    }
+}
+
+/// Telemetry handle used to send metrics to the agent
+pub struct TelemetryClient {
+    metrics: Metrics,
+    worker: TelemetryWorkerHandle,
+    handle: JoinHandle<()>,
+}
+
+impl TelemetryClient {
+    async fn add_point(&self, value: f64, key: ContextKey, tags: Vec<Tag>) {
+        // TODO: map error.
+        let _ = self
+            .worker
+            .send_msg(TelemetryActions::AddPoint((value, key, tags)))
+            .await;
+    }
+
+    /// Sends metrics to the agent using a telemetry worker handle.
+    ///
+    /// # Arguments:
+    ///
+    /// * `telemetry_handle`: telemetry worker handle used to enqueue metrics.
+    pub async fn send(&self, data: &SendDataResult) {
+        let mut futures = Vec::new();
+        if data.requests_count > 0 {
+            if let Some(key) = self.metrics.get(API_REQUEST_STR) {
+                futures.push(self.add_point(data.requests_count as f64, *key, vec![]));
+            }
+        }
+        if data.errors_network > 0 {
+            if let Some(key) = self.metrics.get(API_ERRORS_STR) {
+                futures.push(self.add_point(
+                    data.errors_network as f64,
+                    *key,
+                    vec![tag!("type", "network")],
```

Comment:
if you are hardcoding tags, it'll be more efficient to add them when registering the metric contexts instead if on every points

Reply:
+1. Perhaps we could also use consts if we're hardcoding strs?